packages/utilities/src/classNamesFunction.ts

```diff
@@ -12,29 +25,114 @@ export interface IClassNamesFunctionOptions {
 /**
  * Creates a getClassNames function which calls getStyles given the props, and injects them
  * into mergeStyleSets.
+ *
+ * Note that the props you pass in on every render should be in the same order and
+ * immutable (numbers, strings, and booleans). This will allow the results to be memoized. Violating
+ * these will cause extra recalcs to occur.
  */
 export function classNamesFunction<TStyleProps extends {}, TStyleSet extends IStyleSet<TStyleSet>>(
   options: IClassNamesFunctionOptions = {}
 ): (getStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined, styleProps?: TStyleProps) => IProcessedStyleSet<TStyleSet> {
-  // TODO: memoize.
+  // We build a trie where each node is a Map. The map entry key represents an argument
+  // value, and the entry value is another node (Map). Each node has a `__retval__`
+  // property which is used to hold the cached response.
+
+  // To derive the response, we can simply ensure the arguments are added or already
+  // exist in the trie. At the last node, if there is a `__retval__` we return that. Otherwise
```

Comment:
Nit: tree

Reply:
No, it's actually a [trie](https://en.wikipedia.org/wiki/Trie). 😄